diff --git a/klippy/chelper/serialqueue.c b/klippy/chelper/serialqueue.c
index e6810933..67374378 100644
--- a/klippy/chelper/serialqueue.c
+++ b/klippy/chelper/serialqueue.c
@@ -84,6 +84,7 @@ struct serialqueue {
 
 #define SQT_UART 'u'
 #define SQT_CAN 'c'
+#define SQT_SOCKET 's'
 #define SQT_DEBUGFILE 'f'
 
 #define MIN_RTO 0.025
diff --git a/klippy/mcu.py b/klippy/mcu.py
index 68179bdd..25360e73 100644
--- a/klippy/mcu.py
+++ b/klippy/mcu.py
@@ -567,12 +567,17 @@ class MCU:
         self._serial = serialhdl.SerialReader(self._reactor, warn_prefix=wp)
         self._baud = 0
         self._canbus_iface = None
+        self._tcp_port = None
         canbus_uuid = config.get('canbus_uuid', None)
+        host_dev = config.get('host', None)
         if canbus_uuid is not None:
             self._serialport = canbus_uuid
             self._canbus_iface = config.get('canbus_interface', 'can0')
             cbid = self._printer.load_object(config, 'canbus_ids')
             cbid.add_uuid(config, canbus_uuid, self._canbus_iface)
+        elif host_dev is not None:
+            self._serialport = host_dev
+            self._tcp_port = config.getint('port', 8888)
         else:
             self._serialport = config.get('serial')
             if not (self._serialport.startswith("/dev/rpmsg_")
@@ -581,7 +586,7 @@ class MCU:
         # Restarts
         restart_methods = [None, 'arduino', 'cheetah', 'command', 'rpi_usb']
         self._restart_method = 'command'
-        if self._baud:
+        if self._baud or self._tcp_port:
             rmethods = {m: m for m in restart_methods}
             self._restart_method = config.getchoice('restart_method',
                                                     rmethods, None)
@@ -793,6 +798,8 @@ class MCU:
                     nodeid = cbid.get_nodeid(self._serialport)
                     self._serial.connect_canbus(self._serialport, nodeid,
                                                 self._canbus_iface)
+                elif self._tcp_port:
+                    self._serial.connect_net(self._serialport, self._tcp_port)
                 elif self._baud:
                     # Cheetah boards require RTS to be deasserted
                     # else a reset will trigger the built-in bootloader.
diff --git a/klippy/serialhdl.py b/klippy/serialhdl.py
index f3c4ed99..6252ba00 100644
--- a/klippy/serialhdl.py
+++ b/klippy/serialhdl.py
@@ -3,7 +3,7 @@
 # Copyright (C) 2016-2021  Kevin O'Connor <kevin@koconnor.net>
 #
 # This file may be distributed under the terms of the GNU GPLv3 license.
-import logging, threading, os
+import logging, threading, os, socket
 import serial
 
 import msgproto, chelper, util
@@ -195,6 +195,30 @@ class SerialReader:
             ret = self._start_session(serial_dev)
             if ret:
                 break
+    def connect_net(self, netdevice, tcpport):
+        # Initial connection
+        logging.info("%sStarting network connect", self.warn_prefix)
+        start_time = self.reactor.monotonic()
+        while 1:
+            if self.reactor.monotonic() > start_time + 90.:
+                self._error("Unable to connect")
+            try:
+                serial_dev = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+                serial_dev.setsockopt(socket.IPPROTO_TCP, socket.TCP_USER_TIMEOUT, 100)
+                serial_dev.setsockopt(socket.SOL_SOCKET,  socket.SO_KEEPALIVE, 1)
+                serial_dev.setsockopt(socket.IPPROTO_TCP, socket.TCP_KEEPCNT, 3)
+                serial_dev.setsockopt(socket.IPPROTO_TCP, socket.TCP_KEEPIDLE, 1)
+                serial_dev.setsockopt(socket.IPPROTO_TCP, socket.TCP_KEEPINTVL, 1)
+                serial_dev.connect((netdevice, tcpport))
+            except (OSError, IOError) as e:
+                logging.warn("%sUnable to connect to host: %s",
+                             self.warn_prefix, e)
+                self.reactor.pause(self.reactor.monotonic() + 5.)
+                continue
+            #stk500v2_leave(serial_dev, self.reactor)
+            ret = self._start_session(serial_dev, b's')
+            if ret:
+                break
     def connect_file(self, debugoutput, dictionary, pace=False):
         self.serial_dev = debugoutput
         self.msgparser.process_identify(dictionary, decompress=False)
diff --git a/src/stm32/Kconfig b/src/stm32/Kconfig
index e3961176..66030d99 100644
--- a/src/stm32/Kconfig
+++ b/src/stm32/Kconfig
@@ -517,4 +517,8 @@ config STM32_CANBUS_PC2_PC3
     bool
     default y if STM32_MMENU_CANBUS_PC2_PC3 || STM32_CMENU_CANBUS_PC2_PC3
 
+config STM32_ENABLE_WIFI
+    bool "Enable support MKS Robin WiFi module"
+    default n
+
 endif
diff --git a/src/stm32/Makefile b/src/stm32/Makefile
index 18af2e9d..cf0ede35 100644
--- a/src/stm32/Makefile
+++ b/src/stm32/Makefile
@@ -29,6 +29,7 @@ CFLAGS-$(CONFIG_MACH_STM32G0) += -mcpu=cortex-m0plus -Ilib/stm32g0/include
 CFLAGS-$(CONFIG_MACH_STM32G4) += -mcpu=cortex-m4 -Ilib/stm32g4/include
 CFLAGS-$(CONFIG_MACH_STM32H7) += -mcpu=cortex-m7 -Ilib/stm32h7/include
 CFLAGS-$(CONFIG_MACH_STM32L4) += -mcpu=cortex-m4 -Ilib/stm32l4/include
+CFLAGS-$(CONFIG_ENABLE_WIFI)  += -DSTM32_ENABLE_WIFI
 CFLAGS += $(CFLAGS-y) -D$(MCU_UPPER) -mthumb -Ilib/cmsis-core -Ilib/fast-hash
 
 CFLAGS_klipper.elf += --specs=nano.specs --specs=nosys.specs
diff --git a/src/stm32/serial.c b/src/stm32/serial.c
index 19aa048e..0f7c05b3 100644
--- a/src/stm32/serial.c
+++ b/src/stm32/serial.c
@@ -7,6 +7,8 @@
 #include "autoconf.h" // CONFIG_SERIAL_BAUD
 #include "board/armcm_boot.h" // armcm_enable_irq
 #include "board/serial_irq.h" // serial_rx_byte
+#include "board/gpio.h"
+#include "generic/armcm_timer.h" // udelay
 #include "command.h" // DECL_CONSTANT_STR
 #include "internal.h" // enable_pclock
 #include "sched.h" // DECL_INIT
@@ -81,6 +83,23 @@ serial_enable_tx_irq(void)
 void
 serial_init(void)
 {
+#if defined(STM32_ENABLE_WIFI)
+    struct gpio_out MKS_WIFI_IO0 = gpio_out_setup(GPIO('A',8), 0);
+    gpio_out_write(MKS_WIFI_IO0, 1);
+
+    struct gpio_out MKS_WIFI_IO4 = gpio_out_setup(GPIO('C', 7), 0);
+    gpio_out_write(MKS_WIFI_IO4, 1);
+
+    struct gpio_out MKS_WIFI_IO_RST = gpio_out_setup(GPIO('A', 5), 0);
+    gpio_out_write(MKS_WIFI_IO_RST, 0);
+
+    udelay(200 * 1000);
+    gpio_out_write(MKS_WIFI_IO_RST, 1);
+
+    udelay(200 * 1000);
+    gpio_out_write(MKS_WIFI_IO4, 0);
+#endif
+
     enable_pclock((uint32_t)USARTx);
 
     uint32_t pclk = get_pclock_frequency((uint32_t)USARTx);
